File: packages/protocol-types/src/game-state.ts
/**
 * RoundState and GameInstance types.
 *
 * Per `spec/state-machine.md` §5 (the canonical state-class set) and the
 * `RoundState` / `GameInstance` shapes used by the test vectors.
 */

import type {
  BlockHeight,
  GameId,
  Hash256,
  RoundNumber,
  RuleSetHash,
  Satoshis,
  Seat,
} from './primitives.js';
import type { ActionType } from './actions.js';
import type { RevealedCard } from './cards.js';
import type { PlayerState } from './player.js';

/** State classes per `spec/state-machine.md` §3. */
export type StateClass =
  | 'S0_TABLE_OPEN'
  | 'S1_SEAT_OPEN'
  | 'S2_TABLE_LOCKED'
  | 'S3_ENTROPY_COMMIT_WINDOW'
  | 'S4_ENTROPY_REVEAL_WINDOW'
  | 'S5_DECK_COMMITTED'
  | 'S6_CARD_REVEAL_FIRST'
  | 'S7_CARD_REVEAL_SECOND'
  | 'S8_BET_DECISION'
  | 'S9_CARD_REVEAL_THIRD'
  | 'S10_SETTLED_ROUND'
  | 'S11_ROTATE_TURN'
  | 'S12_TABLE_CLOSE'
  | 'RECOVERED';

/** Lifecycle status of a {@link GameInstance}. */
export type LifecycleStatus =
  | 'open'
  | 'locked'
  | 'shuffling'
  | 'dealing'
  | 'active'
  | 'settling'
  | 'closed'
  | 'recovered';

/**
 * The canonical state object committed at every node of the protocol's
 * transaction tree. Two honest implementations MUST compute byte-identical
 * encodings of this record for the same logical input.
 */
export interface RoundState {
  readonly state_class: StateClass;
  readonly game_id: GameId;
  readonly rule_set_hash: RuleSetHash;
  readonly round_number: RoundNumber;
  /** Acting player's seat; null for quorum-driven or deterministic states. */
  readonly acting_player_seat: Seat | null;
  readonly players: readonly PlayerState[];
  readonly pot_value: Satoshis;
  readonly visible_cards: readonly RevealedCard[];
  /** References (commitment hashes) to deck positions not yet opened. */
  readonly hidden_commitment_refs: readonly Hash256[];
  /** Actions legal from this state. */
  readonly allowed_actions: readonly ActionType[];
  /** Block height after which the timeout branch becomes spendable; null = no timeout. */
  readonly decision_deadline_block_height: BlockHeight | null;
  /** Absolute block height after which the recovery branch becomes spendable. */
  readonly recovery_deadline_block_height: BlockHeight | null;
  /** Set of canonical-template hashes that name the legal successors. */
  readonly successor_template_hashes: readonly Hash256[];
  /** Hash of the parent state in the transition tree; null at the root. */
  readonly prior_state_hash: Hash256 | null;
  /**
   * State hash computed by domain-separated SHA-256 over the canonical
   * encoding of this record with this field set to 64 zero chars.
   */
  readonly state_hash: Hash256;
}

/**
 * Top-level session object. Cross-state aggregate; useful for indexers
 * and the client lobby, but the state engine itself operates on
 * {@link RoundState}.
 */
export interface GameInstance {
  readonly game_id: GameId;
  readonly rule_set_hash: RuleSetHash;
  /** Stable ordering: players in canonical seat order. */
  readonly player_list: readonly PlayerState[];
  /** Turn order (seat permutation); turn_order[i] is the seat of the i-th turn. */
  readonly turn_order: readonly Seat[];
  readonly current_round_number: RoundNumber;
  readonly current_state_hash: Hash256;
  readonly lifecycle_status: LifecycleStatus;
}

File: apps/relay-go/pkg/types/game_state.go
package types

// StateClass enumerates every node class of the state machine.
type StateClass string

const (
	StateTableOpen           StateClass = "S0_TABLE_OPEN"
	StateSeatOpen            StateClass = "S1_SEAT_OPEN"
	StateTableLocked         StateClass = "S2_TABLE_LOCKED"
	StateEntropyCommit       StateClass = "S3_ENTROPY_COMMIT_WINDOW"
	StateEntropyReveal       StateClass = "S4_ENTROPY_REVEAL_WINDOW"
	StateDeckCommitted       StateClass = "S5_DECK_COMMITTED"
	StateCardRevealFirst     StateClass = "S6_CARD_REVEAL_FIRST"
	StateCardRevealSecond    StateClass = "S7_CARD_REVEAL_SECOND"
	StateBetDecision         StateClass = "S8_BET_DECISION"
	StateCardRevealThird     StateClass = "S9_CARD_REVEAL_THIRD"
	StateSettledRound        StateClass = "S10_SETTLED_ROUND"
	StateRotateTurn          StateClass = "S11_ROTATE_TURN"
	StateTableClose          StateClass = "S12_TABLE_CLOSE"
	StateRecovered           StateClass = "RECOVERED"
)

// LifecycleStatus is the GameInstance top-level state. Useful for
// indexers; the state engine itself works on RoundState.
type LifecycleStatus string

const (
	LifecycleOpen       LifecycleStatus = "open"
	LifecycleLocked     LifecycleStatus = "locked"
	LifecycleShuffling  LifecycleStatus = "shuffling"
	LifecycleDealing    LifecycleStatus = "dealing"
	LifecycleActive     LifecycleStatus = "active"
	LifecycleSettling   LifecycleStatus = "settling"
	LifecycleClosed     LifecycleStatus = "closed"
	LifecycleRecovered  LifecycleStatus = "recovered"
)

// RoundState is the canonical state object committed at every node of
// the protocol's transaction tree. Two honest engines must compute
// byte-identical encodings for the same logical input.
type RoundState struct {
	StateClass                  StateClass     `json:"state_class"`
	GameId                      GameId         `json:"game_id"`
	RuleSetHash                 RuleSetHash    `json:"rule_set_hash,omitempty"`
	RoundNumber                 RoundNumber    `json:"round_number"`
	ActingPlayerSeat            *Seat          `json:"acting_player_seat"`
	Players                     []PlayerState  `json:"players"`
	PotValue                    Satoshis       `json:"pot_value"`
	VisibleCards                []RevealedCard `json:"visible_cards"`
	HiddenCommitmentRefs        []Hash256      `json:"hidden_commitment_refs"`
	AllowedActions              []ActionType   `json:"allowed_actions"`
	DecisionDeadlineBlockHeight *BlockHeight   `json:"decision_deadline_block_height"`
	RecoveryDeadlineBlockHeight *BlockHeight   `json:"recovery_deadline_block_height"`
	SuccessorTemplateHashes     []Hash256      `json:"successor_template_hashes"`
	PriorStateHash              *Hash256       `json:"prior_state_hash"`
	StateHash                   Hash256        `json:"state_hash"`
}

// GameInstance is a session-level summary. Indexers use this; the
// engine operates on RoundState.
type GameInstance struct {
	GameId             GameId           `json:"game_id"`
	RuleSetHash        RuleSetHash      `json:"rule_set_hash"`
	PlayerList         []PlayerState    `json:"player_list"`
	TurnOrder          []Seat           `json:"turn_order"`
	CurrentRoundNumber RoundNumber      `json:"current_round_number"`
	CurrentStateHash   Hash256          `json:"current_state_hash"`
	LifecycleStatus    LifecycleStatus  `json:"lifecycle_status"`
}
